Mage.Sets/src/mage/cards/t/TheFifteenthDoctor.java

```diff
@@ -0,0 +1,160 @@
+        return new TheFifteenthDoctor(this);
+    }
+}
+
+class TheFifteenthDoctorEffect extends OneShotEffect {
+
+    public static final FilterCard filter = new FilterArtifactCard("an artifact card with mana value 2 or 3");
+
+    static {
+        filter.add(Predicates.or(
+                new ManaValuePredicate(ComparisonType.EQUAL_TO, 2),
+                new ManaValuePredicate(ComparisonType.EQUAL_TO, 3)
+        ));
+    }
+
+    TheFifteenthDoctorEffect() {
+        super(Outcome.Benefit);
+        staticText = "mill three cards. You may put an artifact card " +
+                "with mana value 2 or 3 from among them into your hand.";
+    }
+
+    private TheFifteenthDoctorEffect(final TheFifteenthDoctorEffect effect) { super(effect); }
+
+    @Override
+    public TheFifteenthDoctorEffect copy() { return new TheFifteenthDoctorEffect(this); }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        if (controller == null) {
+            return false;
+        }
+        Cards cards = controller.millCards(3, source, game);
+        String message = "Put " + filter.getMessage() +
+                " from among the milled cards into your hand?";
+        if (cards.isEmpty()
+                || !controller.chooseUse(outcome, message, source, game)) {
+            return true;
+        }
+        TargetCard target = new TargetCard(Zone.GRAVEYARD, filter);
+        if (controller.choose(outcome, cards, target, source, game)) {
+            Card card = game.getCard(target.getFirstTarget());
+            if (card != null) {
+                return controller.moveCards(card, Zone.HAND, source, game);
+            }
+        }
+        return false;
+    }
+}
+
+class TheFifteenthDoctorSpellWatcher extends Watcher {
+    private final Map<UUID, Integer> nonartifactSpells;
+
+    public TheFifteenthDoctorSpellWatcher() {
+        super(WatcherScope.GAME);
+        nonartifactSpells = new HashMap<>();
+    }
+
+    @Override
+    public void watch(GameEvent event, Game game) {
+        if (event.getType() == GameEvent.EventType.SPELL_CAST) {
+            Spell spell = (Spell) game.getObject(event.getTargetId());
+            if (spell != null && !spell.isArtifact(game)) {
+                nonartifactSpells.put(event.getPlayerId(), nonartifactSpellsCastThisTurn(event.getPlayerId()) + 1);
+            }
+        }
+    }
+
+    public int nonartifactSpellsCastThisTurn(UUID playerId) { return nonartifactSpells.getOrDefault(playerId, 0); }
+
+    @Override
+    public void reset() {
+        super.reset();
+        nonartifactSpells.clear();
+    }
+}
+
+enum TheFifteenthDoctorCastSpellPredicate implements ObjectSourcePlayerPredicate<Card> {
+    instance;
+
+    @Override
+    public boolean apply(ObjectSourcePlayer<Card> input, Game game) {
+        if (input.getObject() != null && !input.getObject().isArtifact(game)) {
```

Comment:
No need in object filter in predicates. Player and object filled all the time. Same for source param (there are possible rare use cases with empty source but it not related to predicates).